<?xml version="1.0"?>
<robot name="uirover"
    xmlns:xacro="http://www.ros.org/wiki/xacro">
    <xacro:arg name="hw_type" default="target"/>
    <xacro:property name='hw_type' value="$(arg hw_type)"/>

    <gazebo>
        <plugin filename="gz_ros2_control-system" name="gz_ros2_control::GazeboSimROS2ControlPlugin">
            <ros>
                <namespace>/uirover/control</namespace>
            </ros>
            <parameters>$(find uirover_controllers)/config/controllers.yaml</parameters>
        </plugin>
        <plugin filename="gz-sim-sensors-system" name="gz::sim::systems::Sensors">
            <render_engine>ogre2</render_engine>
        </plugin>
    </gazebo>

    <ros2_control name="wheels_hardware" type="system">
        <hardware>
            <xacro:if value="${hw_type=='gazebo'}">
                <plugin>gz_ros2_control/GazeboSimSystem</plugin>
            </xacro:if>
            <xacro:if value="${hw_type=='target'}">
                <plugin>uirover_hardware/UiroverHardwareInterface</plugin>
            </xacro:if>
            <xacro:if value="${hw_type=='mock'}">
                <plugin>mock_components/GenericSystem</plugin>
            </xacro:if>
        </hardware>

        <joint name="fl_wheel_joint">
            <command_interface name="velocity">
                <param name="min">-1</param>
                <param name="max">1</param>
            </command_interface>
            <state_interface name="velocity" />
            <state_interface name="position" />
        </joint>
        <joint name="fr_wheel_joint">
            <command_interface name="velocity">
                <param name="min">-1</param>
                <param name="max">1</param>
            </command_interface>
            <state_interface name="velocity" />
            <state_interface name="position" />
        </joint>
        <joint name="bl_wheel_joint">
            <command_interface name="velocity">
                <param name="min">-1</param>
                <param name="max">1</param>
            </command_interface>
            <state_interface name="velocity" />
            <state_interface name="position" />
        </joint>
        <joint name="br_wheel_joint">
            <command_interface name="velocity">
                <param name="min">-1</param>
                <param name="max">1</param>
            </command_interface>
            <state_interface name="velocity" />
            <state_interface name="position" />
        </joint>
        <joint name="ml_wheel_joint">
            <command_interface name="velocity">
                <param name="min">-1</param>
                <param name="max">1</param>
            </command_interface>
            <state_interface name="velocity" />
            <state_interface name="position" />
        </joint>
        <joint name="mr_wheel_joint">
            <command_interface name="velocity">
                <param name="min">-1</param>
                <param name="max">1</param>
            </command_interface>
            <state_interface name="velocity" />
            <state_interface name="position" />
        </joint>
    </ros2_control>


    <ros2_control name="arm_hardware" type="system" params="hw_type:='target'">
        <hardware>
            <!-- Hardware plugin for Gazebo Sim -->
            <xacro:if value="${hw_type=='gazebo'}">
                <plugin>gz_ros2_control/GazeboSimSystem</plugin>
            </xacro:if>

            <!-- Real Hardware -->
            <xacro:if value="${hw_type=='target'}">
                <plugin>uirover_hardware/UiroverHardwareInterface</plugin>
            </xacro:if>

            <!-- All state interfaces match coresponding command interface exactly -->
            <xacro:if value="${hw_type=='mock'}">
                <plugin>mock_components/GenericSystem</plugin>
            </xacro:if>
        </hardware>

        <joint name="arm_j0">
            <command_interface name="position" />
            <state_interface name="position">
                <param name="initial_value">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <param name="odrive_serial_number">C0FFEE00</param>
        </joint>
        <joint name="arm_j1">
            <command_interface name="position" />
            <state_interface name="position">
                <param name="initial_value">0.262</param>
            </state_interface>
            <state_interface name="velocity" />
            <param name="odrive_serial_number">C0FFEE01</param>
        </joint>
        <joint name="arm_j2">
            <command_interface name="position" />
            <state_interface name="position">
                <param name="initial_value">1.745</param>
            </state_interface>
            <state_interface name="velocity" />
            <param name="odrive_serial_number">C0FFEE02</param>
        </joint>
        <joint name="arm_j3">
            <command_interface name="position" />
            <state_interface name="position">
                <param name="initial_value">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <param name="odrive_serial_number">C0FFEE03</param>
        </joint>
        <joint name="arm_j4">
            <command_interface name="position" />
            <state_interface name="position">
                <param name="initial_value">-0.436</param>
            </state_interface>
            <state_interface name="velocity" />
            <param name="odrive_serial_number">C0FFEE04</param>
        </joint>
        <joint name="arm_j5">
            <command_interface name="position" />
            <state_interface name="position">
                <param name="initial_value">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <param name="odrive_serial_number">C0FFEE05</param>
        </joint>

    </ros2_control>
</robot>